### ISSUE TITLE
gcloud-python-bigtable invalid certs file

### ISSUE THREAD

**Issue by hilld1** [dhermes/gcloud-python-bigtable#21]
I set up everything, I use a JSON key as credentials.

```python
from gcloud_bigtable.client import Client
client = Client(admin = True)
#client.start()
print client.list_zones()
```

When I run this code I get :

```python
gcloud_bigtable/client.py", line 411, in _cluster_stub
    raise ValueError('Client has not been started.')
ValueError: Client has not been started."
```

So I start the client only to get the following: 

```python
gcloud_bigtable/_helpers.py", line 244, in _set_certs
    with open(SSL_CERT_FILE, mode='rb') as file_obj:
IOError: [Errno 2] No such file or directory: '/etc/ssl/certs/ca-certificates.crt'"
```

All I want to do is access the table in my cluster and add or remove rows, but I am stuck before I even manage to get a cluster.

Can someone explain what I'm missing ?

Thank you!

**Comment by dhermes** [comment]
Thanks a lot @hilld1! This is the most brittle part of the code and I'm not sure how to solve the problem yet. Eseentially, I have hardcoded a system path for an SSL certificates file stored on the system. This is required (as far as I know) for making authenticated requests with gRPC. It'd be really great if @nathanielmanistaatgoogle could shed some light on this?

----

First a question for you: what is your operating system?

----

To see what is happening, look at the [code][1]:

```python
EPOCH = datetime.datetime.utcfromtimestamp(0).replace(tzinfo=pytz.utc)
# See https://gist.github.com/dhermes/bbc5b7be1932bfffae77
# for appropriate values on other systems.
SSL_CERT_FILE = '/etc/ssl/certs/ca-certificates.crt'
```

I recommend visiting https://gist.github.com/dhermes/bbc5b7be1932bfffae77 and running `find_ssl_cert.py` on your machine so we can get a sense of where else this path might be going.

[1]: https://github.com/dhermes/gcloud-python-bigtable/blob/0f26ad2b3a6eb7f9e988d7bb5676e589eb7c6a49/gcloud_bigtable/_helpers.py#L47-L50

**Comment by hilld1** [comment]
Hello,
I am running on OSX 

I ran find_ssl_cert.py:
 python find_ssl_cert.py 
SSL_CERT_FILE matches:
    /usr/local/etc/openssl/cert.pem
No SSL_CERT_DIR matches found
Linux distro:

**Comment by dhermes** [comment]
Thanks @hilld1 that explains it. I'll put in a hack ASAP and ping this thread once it's in.

**Comment by dhermes** [comment]
@hilld1 FYI I added support in d16fa35c4070445d38a5103914c1c207487eb7ed

You should update your version of this library and see if it works.

**Comment by hilld1** [comment]
I updated my version and ran the same code as before, I got the following: 

Traceback (most recent call last):
  File ".py", line 7, in <module>
    client.start()
  File "../gcloud-python-bigtable-master/gcloud_bigtable/client.py", line 508, in start
    self._data_stub_internal = self._make_data_stub()
  File "../gcloud-python-bigtable-master/gcloud_bigtable/client.py", line 454, in _make_data_stub
    DATA_API_HOST, DATA_API_PORT)
  File "../gcloud-python-bigtable-master/gcloud_bigtable/_helpers.py", line 270, in make_stub
    root_certificates = get_certs()
  File "../gcloud-python-bigtable-master/gcloud_bigtable/_helpers.py", line 244, in get_certs
    set_certs(reset=False)
  File "../gcloud-python-bigtable-master/gcloud_bigtable/_helpers.py", line 232, in set_certs
    _set_certs()
  File "../gcloud-python-bigtable-master/gcloud_bigtable/_helpers.py", line 218, in _set_certs
    with open(SSL_CERT_FILE, mode='rb') as file_obj:
IOError: [Errno 2] No such file or directory: '/etc/ssl/certs/ca-certificates.crt'

**Comment by dhermes** [comment]
@hilld1 I [think][1] your upgrade didn't go smoothly. Try

```
pip install --ignore-installed git+https://github.com/dhermes/gcloud-python-bigtable#egg=gcloud-bigtable
```

(I'm not sure why but the `--force-reinstall` flag doesn't do the trick.)

[1]: https://github.com/dhermes/gcloud-python-bigtable/commit/d16fa35c4070445d38a5103914c1c207487eb7ed#commitcomment-14737277

**Comment by dhermes** [comment]
Haha nope! I made an error, I put the Linux version under Mac and the Mac version under Linux!

**Comment by dhermes** [comment]
@hilld1 Swapped the filenames to be correct in c95d0cfacbe7e64e760363f3d5498b4c31e7a270. D'oh!

**Comment by hilld1** [comment]
I ran it and my terminal printed an error and got stuck.
```
File "..Test.py", line 8, in <module>
    client.list_clusters()
  File "../gcloud-python-bigtable-master/gcloud_bigtable/client.py", line 603, in list_clusters
    request_pb, timeout_seconds)
  File "/Library/Python/2.7/site-packages/grpc/framework/crust/implementations.py", line 73, in __call__
    protocol_options, metadata, request)
  File "/Library/Python/2.7/site-packages/grpc/framework/crust/_calls.py", line 109, in blocking_unary_unary
    return next(rendezvous)
  File "/Library/Python/2.7/site-packages/grpc/framework/crust/_control.py", line 412, in next
    raise self._termination.abortion_error
grpc.framework.interfaces.face.face.NetworkError
```

**Comment by dhermes** [comment]
That's progress! Two thing

1.  You said "I ran it". What is **it**?
1.  The `NetworkError` [actually contains][1] an error code and an error message. You should catch it

    ```python
    from grpc.framework.interfaces.face.face import NetworkError
    try:
        do_something()
    except NetworkError as exc:
        print exc.code
        print exc.details
    ```

    to see what error your API requests is causing.

1.  The fact that your program hangs at the end is a known issue. @nathanielmanistaatgoogle mentioned that it would be enough to remove the stubs from the variable they were assigned to but I think it may be necessary to `del self._attached_stub`

------------

I may end up wrapping the API methods to catch a `NetworkError` so I can give a more helpful message.

[1]: https://github.com/grpc/grpc/issues/2611#issuecomment-159721031

**Comment by dhermes** [comment]
This is fixed by `0.13.0`
